File: smoke_tests/typescript/insecure_cors.ts
// Insecure CORS Test Cases

import { Request, Response } from 'express';

// Test 1: Wildcard CORS with credentials
function setWildcardCORS(req: Request, res: Response): void {
    // VULNERABLE: Wildcard origin with credentials is not allowed
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Credentials', 'true');
}

// Test 2: Reflecting origin without validation
function reflectOrigin(req: Request, res: Response): void {
    const origin = req.headers.origin;
    // VULNERABLE: Reflecting origin without validation
    res.setHeader('Access-Control-Allow-Origin', origin || '');
    res.setHeader('Access-Control-Allow-Credentials', 'true');
}

// Test 3: Null origin allowed
function allowNullOrigin(req: Request, res: Response): void {
    // VULNERABLE: null origin can be exploited
    res.setHeader('Access-Control-Allow-Origin', 'null');
    res.setHeader('Access-Control-Allow-Credentials', 'true');
}

// Test 4: Overly permissive CORS headers
function setPermissiveCORS(req: Request, res: Response): void {
    // VULNERABLE: Allowing all origins without proper validation
    res.setHeader('Access-Control-Allow-Origin', req.headers.origin || '*');
    res.setHeader('Access-Control-Allow-Methods', '*');
    res.setHeader('Access-Control-Allow-Headers', '*');
}

// Test 5: Trusting subdomain without proper validation
function trustSubdomainWithRegex(req: Request, res: Response): void {
    const origin = req.headers.origin || '';
    // VULNERABLE: Weak regex can be bypassed (e.g., evilexample.com)
    if (origin.match(/example\.com/)) {
        res.setHeader('Access-Control-Allow-Origin', origin);
        res.setHeader('Access-Control-Allow-Credentials', 'true');
    }
}

File: smoke_tests/rust/insecure_cors.rs
// Insecure CORS Test Cases

use actix_web::{HttpRequest, HttpResponse};

// Test 1: Wildcard CORS with credentials
fn set_wildcard_cors() -> HttpResponse {
    // VULNERABLE: Wildcard origin with credentials is not allowed
    HttpResponse::Ok()
        .append_header(("Access-Control-Allow-Origin", "*"))
        .append_header(("Access-Control-Allow-Credentials", "true"))
        .finish()
}

// Test 2: Reflecting origin without validation
fn reflect_origin(req: HttpRequest) -> HttpResponse {
    let origin = req.headers()
        .get("Origin")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");

    // VULNERABLE: Reflecting origin without validation
    HttpResponse::Ok()
        .append_header(("Access-Control-Allow-Origin", origin))
        .append_header(("Access-Control-Allow-Credentials", "true"))
        .finish()
}

// Test 3: Null origin allowed
fn allow_null_origin() -> HttpResponse {
    // VULNERABLE: null origin can be exploited
    HttpResponse::Ok()
        .append_header(("Access-Control-Allow-Origin", "null"))
        .append_header(("Access-Control-Allow-Credentials", "true"))
        .finish()
}

// Test 4: Overly permissive CORS headers
fn set_permissive_cors(req: HttpRequest) -> HttpResponse {
    let origin = req.headers()
        .get("Origin")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("*");

    // VULNERABLE: Allowing all origins without proper validation
    HttpResponse::Ok()
        .append_header(("Access-Control-Allow-Origin", origin))
        .append_header(("Access-Control-Allow-Methods", "*"))
        .append_header(("Access-Control-Allow-Headers", "*"))
        .finish()
}

// Test 5: Weak subdomain validation
fn trust_subdomain_with_regex(req: HttpRequest) -> HttpResponse {
    let origin = req.headers()
        .get("Origin")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");

    // VULNERABLE: Weak regex can be bypassed (e.g., evilexample.com)
    if origin.contains("example.com") {
        return HttpResponse::Ok()
            .append_header(("Access-Control-Allow-Origin", origin))
            .append_header(("Access-Control-Allow-Credentials", "true"))
            .finish();
    }

    HttpResponse::Ok().finish()
}
